### DISCUSSION TITLE
Improve "Extending TinaMarkdown" DX

### DISCUSSION THREAD

**Discussion by jamespohalloran** [tinacms/tinacms#3227] [Ideas]
I stumbled a bit when trying to extend TinaMarkdown recently, I think a few things could be done to improve the experience.

Here's a few ideas I have. Would love to hear any other ideas, or if any of this resonates!

### Loose Typing

The rich-text field has a set of `templates`, each with `fields`. 

TinaMarkdown takes a "components" property, which usually needs to match the rich-text field's `templates`. 

It would be if there was better typing for the props on TinaMarkdown "component". 

<img width="614" alt="Screen Shot 2022-09-23 at 12 41 06 PM" src="https://user-images.githubusercontent.com/3323181/192002794-5fb41ccc-084e-4d66-8885-f0c25f2cccf3.png">

^Digging through props like this can feel a bit like I'm off-roading. 

### Unclear how to override built-in components

Certain built-in markdown elements can be overridden (E.g `h1`, `h2`, `a`). We should show in the docs how to do this, and what built-in elements are available to override. 

### Handling nested rich-text can be difficult

Let's say you want to give each of your headings an anchor tag. 

E.g
```md
## Hello World
```
would translate to <h1 id="hello-world"></h1> 

It's not too difficult if the value is always a single text child, but this quickly gets more complex if you're trying to handle a node with nested rich-text, or multiple children.

```md
## Hello `world` with a [link]("foobar.blah")
```

I'm not 100% on the best way to handle this, but I feel that exposing some helpers to "stringify" a rich-text node would be helpful.

**Comment by jeffsee55** [comment]
> Loose Typing

I'm not sure what you mean here, in the example screenshot TOCTitle1's `props` is implicitly `any`. We provide decent types as far as I know, we could include types from a given `rich-text`'s `templates` but that would need to be part of our autogenerated types (would love to do this)

> Handling nested rich-text can be difficult

There are some handy helpers we could expose, in this example you'd want a helper to "unwrap" the node so it doesn't have any links, bold, etc. That logic is mostly in the `@tinacms/mdx` package but depending on what we want to use it for it might not be too hard to pull some of it out.

**Reply by jamespohalloran** [reply]
> > Loose Typing
> 
> I'm not sure what you mean here, in the example screenshot TOCTitle1's `props` is implicitly `any`. We provide decent types as far as I know, we could include types from a given `rich-text`'s `templates` but that would need to be part of our autogenerated types (would love to do this)

Here's a bit more detailed demo of what I mean: https://www.loom.com/share/fca5f3a19db94d1caa7cb30c44f4832c
This is using the Tina Cloud starter for reference.

**Reply by jeffsee55** [reply]
I see, yeah we expose a [`Components` generic](https://github.com/tinacms/tina-cloud-starter/blob/main/components/posts/post.tsx#L22) to help a bit but we could probably improve that so it's possible to pass it right inline. 

The other suggestion about it knowing about your shape would be awesome, I've thought about this before and I think we'd just need to annotate it as a generic and then expose a helper, only weird thing is that helper needs to be unique for each `rich-text` field:

```ts
import { TinaMarkdown } from "tinacms/dist/rich-text";
import type { RichTextFields } from "../.tina/__generated__/types";

// `[post.body]` is `rich-text` field with a Cta template
<TinaMarkdown<RichTextFields['post.body']> components={{Cta: (props) => <Cta {...props} />}} />
```

About the `body` being `any`. I believe that's because the autogenerated SDK doesn't know what to do with the `JSON` primitive. We could improve this with the more advanced one we've discussed.

**Reply by L-A** [reply]
> There are some handy helpers we could expose, in this example you'd want a helper to "unwrap" the node so it doesn't have any links, bold, etc.

Chiming in that I'm currently working towards this, to give editors the ability to insert an automatically populated table of contents in an article. Surprisingly, providing the whole markdown body to a custom `TOC` component is the easy part, done with a wrapper that adds the `body` AST to its `props`.

However, in trying to give each title a URL-friendly `id` value, I need to not _render_ the children directly (by providing them as a `prop` to my custom component), but instead parse them into a flat string. This is frustrated by the fact that the TOC can work _only_ with the AST, but then the `h1`, `h2`, `h3` components can work _only_ with their React `<TinaMarkdown>` children props. This means the TOC needs to work out `id` strings from a _very_ different source of truth than the headings themselves.
